src/scratch_comment_icon.js

```diff
@@ -0,0 +1,221 @@
+          oldBubbleLocation,
+          newBubbleLocation,
+        ),
+      );
+    }
+  }
+
+  setText(text) {
+    this.commentBubble?.setText(text);
+  }
+
+  getText() {
+    return this.commentBubble?.getText() ?? '';
+  }
+
+  onTextChanged(oldText, newText) {
+    Blockly.Events.fire(
+      new (Blockly.Events.get(Blockly.Events.BLOCK_CHANGE))(
+        this.sourceBlock,
+        'comment',
+        null,
+        oldText,
+        newText,
+      ),
+    );
+    Blockly.Events.fire(
+      new (Blockly.Events.get('block_comment_change'))(
+        this.commentBubble,
+        oldText,
+        newText,
+      ),
+    );
+  }
+
+  onCollapsed(collapsed) {
+    Blockly.Events.fire(
+      new (Blockly.Events.get('block_comment_collapse'))(
+        this.commentBubble,
+        collapsed,
+      ),
+    );
+  }
+
+  onSizeChanged(oldSize, newSize) {
+    Blockly.Events.fire(
+      new (Blockly.Events.get('block_comment_resize'))(
+        this.commentBubble,
+        oldSize,
+        newSize,
+      ),
+    );
+  }
+
+  setBubbleSize(size) {
+    this.commentBubble?.setSize(size);
+  }
+
+  getBubbleSize() {
+    return this.commentBubble?.getSize() ?? new Blockly.utils.Size(0, 0);
+  }
+
+  setBubbleLocation(newLocation) {
+    const oldLocation = this.getBubbleLocation();
+    this.commentBubble?.moveTo(newLocation);
+    Blockly.Events.fire(
+      new (Blockly.Events.get('block_comment_move'))(
+        this.commentBubble,
+        oldLocation,
+        newLocation,
+      ),
+    );
+  }
+
+  getBubbleLocation() {
+    return this.commentBubble?.getRelativeToSurfaceXY();
+  }
+
+  saveState() {
+    if (this.commentBubble) {
+      const size = this.getBubbleSize();
+      const bubbleLocation = this.commentBubble.getRelativeToSurfaceXY();
+      const delta = Blockly.utils.Coordinate.difference(bubbleLocation, this.workspaceLocation);
+      return {
+        'text': this.getText(),
+        'height': size.height,
+        'width': size.width,
+        'x': delta.x,
+        'y': delta.y,
+        'collapsed': this.commentBubble.isCollapsed(),
+      }
+    }
+    return null;
+  }
+
+  loadState(state) {
+    this.setText(state['text']);
+    this.setBubbleSize(new Blockly.utils.Size(state['width'], state['height']));
+    const delta = new Blockly.utils.Coordinate(state['x'], state['y']);
+    const newBubbleLocation = Blockly.utils.Coordinate.sum(this.workspaceLocation, delta);
+    this.commentBubble.moveTo(newBubbleLocation);
+    this.commentBubble.setCollapsed(state['collapsed']);
+    this.repositionAfterRender = false;
+  }
+
+  bubbleIsVisible() {
+    return true;
+  }
+
+  async setBubbleVisible(visible) {
+    this.commentBubble.setCollapsed(!visible);
+  }
+
+  dispose() {
+    this.commentBubble?.removeTextChangeListener(this.onTextChangedListener);
+    this.commentBubble?.removeSizeChangeListener(this.onSizeChangedListener);
+    this.commentBubble?.removeOnCollapseListener(this.onCollapseListener);
```

Comment:
You shouldn't need to unregister these since JS uses mark-and-sweep. Once the icon is dereferenced, anything that was referenced through the icon will get collected.

Reply:
Done